server/datastore/mysql/apple_mdm.go

```diff
@@ -231,3 +237,284 @@ WHERE
 	}
 	return devices, nil
 }
+
+func (ds *Datastore) IngestMDMAppleDeviceFromCheckin(ctx context.Context, mdmHost fleet.MDMAppleHostDetails) error {
+	return ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+		return ingestMDMAppleDeviceFromCheckinDB(ctx, tx, mdmHost, ds.logger)
+	})
+}
+
+func ingestMDMAppleDeviceFromCheckinDB(
+	ctx context.Context,
+	tx sqlx.ExtContext,
+	mdmHost fleet.MDMAppleHostDetails,
+	logger log.Logger,
+) error {
+	if mdmHost.SerialNumber == "" {
+		return ctxerr.New(ctx, "ingest mdm apple host from checkin expected device serial number but got empty string")
+	}
+	if mdmHost.UDID == "" {
+		return ctxerr.New(ctx, "ingest mdm apple host from checkin expected unique device id but got empty string")
+	}
+
+	stmt := `SELECT id, uuid, hardware_serial FROM hosts WHERE uuid = ? OR hardware_serial = ?`
+
+	var foundHost fleet.Host
+	err := sqlx.GetContext(ctx, tx, &foundHost, stmt, mdmHost.UDID, mdmHost.SerialNumber)
+	switch {
+	case errors.Is(err, sql.ErrNoRows):
+		return insertMDMAppleHostDB(ctx, tx, mdmHost, logger)
+
+	case err != nil:
+		return ctxerr.Wrap(ctx, err, "get mdm apple host by serial number or udid")
+
+	case foundHost.HardwareSerial != mdmHost.SerialNumber || foundHost.UUID != mdmHost.UDID:
+		return updateMDMAppleHostDB(ctx, tx, foundHost.ID, mdmHost)
+
```

Comment:
I think we will always want to update the host if it's not a new record, can we remove this condition?